File: src/SurpherLangMain/Lexer.java
package SurpherLangMain;

import SurpherLangMain.Token.TokenType;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * Lexer for the Surpher language
 */
class Lexer {
    private static final Map<String, TokenType> aKeywords = new HashMap<>();

    /*
      Automatically adds the reserved keywords
     */
    static {
        aKeywords.put("class", TokenType.CLASS);
        aKeywords.put("else", TokenType.ELSE);
        aKeywords.put("false", TokenType.FALSE);
        aKeywords.put("fun", TokenType.FUN);
        aKeywords.put("for", TokenType.FOR);
        aKeywords.put("if", TokenType.IF);
        aKeywords.put("none", TokenType.NONE);
        aKeywords.put("return", TokenType.RETURN);
        aKeywords.put("super", TokenType.SUPER);
        aKeywords.put("self", TokenType.SELF);
        aKeywords.put("true", TokenType.TRUE);
        aKeywords.put("var", TokenType.VAR);
        aKeywords.put("while", TokenType.WHILE);
        aKeywords.put("do", TokenType.DO);
        aKeywords.put("match", TokenType.MATCH);
        aKeywords.put("case", TokenType.CASE);
        aKeywords.put("break", TokenType.BREAK);
        aKeywords.put("continue", TokenType.CONTINUE);
        aKeywords.put("newtype", TokenType.NEWTYPE);
        aKeywords.put("print", TokenType.PRINT);
    }

    private final List<Token> aTokenList = new LinkedList<>();
    private final Predicate<Character> isDigit = pChar -> pChar >= '0' && pChar <= '9';
    private final Predicate<Character> isAlpha = pChar -> (pChar >= 'a' && pChar <= 'z') || (pChar >= 'A' && pChar <= 'Z')
            || pChar == '_';
    private final Predicate<Character> isAlphaNumeric = isAlpha.or(isDigit);
    private String aSource;
    private int aStart = 0;
    private int aCurrent = 0;
    private final Supplier<Character> anyChar = () -> aSource.charAt(aCurrent++);
    private final Supplier<Boolean> isAtEnd = () -> aCurrent >= aSource.length();
    private final Function<Integer, Character> lookAHead = pOffset -> {
        if (isAtEnd.get() || aCurrent + pOffset >= aSource.length()) {
            return '\0';
        } else {
            return aSource.charAt(aCurrent + pOffset);
        }
    };
    private final Function<Character, Boolean> matchNext = pExpectedChar -> {
        if (isAtEnd.get() || aSource.charAt(aCurrent) != pExpectedChar) {
            return false;
        }
        aCurrent++;
        return true;
    };
    private int aLine = 1;
    private final Consumer<TokenType> addType = pType -> addToken(pType, null);

    /**
     * Constructor for the Surpher lexer
     *
     * @param pSource source code
     */
    Lexer(String pSource) {
        aSource = pSource;
    }

    /**
     * Attempts to match a Number type of Surpher
     */
    private void matchNumber() {

        // check if the first char is a digit
        while (isDigit.test(lookAHead.apply(0))) {
            anyChar.get();
        }

        if (lookAHead.apply(0) == '.' && isDigit.test(lookAHead.apply(1))) {
            // case 1: float number
            anyChar.get();
        }  // case 2: int number

        while (isDigit.test(lookAHead.apply(0))) {
            anyChar.get();
        }
        addToken(TokenType.NUMBER, Double.parseDouble(aSource.substring(aStart, aCurrent)));

    }

    /**
     * Attempts to match a Identifier type or a reserved keyword of Surpher
     */
    private void matchIdentifierAndReserved() {
        // check if the leading char is a valid identifier element
        while (isAlphaNumeric.test(lookAHead.apply(0))) {
            anyChar.get();
        }
        String word = aSource.substring(aStart, aCurrent);
        TokenType type = aKeywords.get(word);
        if (type == null) {
            // case 1: identifier
            type = TokenType.IDENTIFIER;
        }
        // case 2: keyword
        addType.accept(type);
    }

    /**
     * Attempts to match a String type of Surpher
     */
    private void matchString() {
        // skip over the string body
        while (!isAtEnd.get() && lookAHead.apply(0) != '"') {
            if (lookAHead.apply(0) == '\n') {
                aLine++;
            }
            anyChar.get();
        }

        // error: unterminated string
        if (isAtEnd.get()) {
            JSurpher.error(aLine, "unterminated string");
            return;
        }

        // skip over the right quotation
        anyChar.get();

        String value = aSource.substring(aStart + 1, aCurrent - 1);
        addToken(TokenType.STRING, value);
    }

    /**
     * Skips over a comment section of Surpher
     */
    private void skipComment() {
        // keep track of the balance
        int flag = 1;

        while (flag != 0 && !isAtEnd.get()) {
            // error: unterminated comment
            if (isAtEnd.get()) {
                JSurpher.error(aLine, "unterminated comment");
                return;
            }

            // update the stack accordingly
            if (matchNext.apply('(') && matchNext.apply('*')) {
                flag++;
            } else if (matchNext.apply('*') && matchNext.apply(')')) {
                flag--;
            }
            if (!isAtEnd.get())
                anyChar.get();
        }
    }

    /**
     * Reads and analyzes the next lexical token in the source code
     */
    private void scanToken() {
        char nextChar = anyChar.get();
        switch (nextChar) {
            case '(' -> {
                if (matchNext.apply('*')) {
                    skipComment();
                } else {
                    addType.accept(TokenType.LEFT_PAREN);
                }
            }
            case ')' -> addType.accept(TokenType.RIGHT_PAREN);
            case '{' -> addType.accept(TokenType.LEFT_BRACE);
            case '}' -> addType.accept(TokenType.RIGHT_BRACE);
            case '[' -> addType.accept(TokenType.LEFT_BRACKET);
            case ']' -> addType.accept(TokenType.RIGHT_BRACKET);
            case ',' -> addType.accept(TokenType.COMMA);
            case '.' -> addType.accept(TokenType.DOT);
            case ';' -> addType.accept(matchNext.apply(';') ? TokenType.DOUBLE_SEMICOLON : TokenType.SINGLE_SEMICOLON);
            case '+' -> addType.accept(matchNext.apply('+') ? TokenType.DOUBLE_PLUS : TokenType.SINGLE_PLUS);
            case '-' -> addType.accept(TokenType.MINUS);
            case '%' -> addType.accept(TokenType.PERCENT);
            case '*' -> addType.accept(TokenType.STAR);
            case '/' -> addType.accept(TokenType.SLASH);
            case '|' -> addType.accept(matchNext.apply('|') ? TokenType.DOUBLE_BAR : TokenType.SINGLE_BAR);
            case '^' -> addType.accept(TokenType.CARET);
            case '&' -> addType.accept(matchNext.apply('&') ? TokenType.DOUBLE_AMPERSAND : TokenType.SINGLE_AMPERSAND);
            case '!' -> addType.accept(matchNext.apply('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
            case '=' -> addType.accept(matchNext.apply('=') ? TokenType.DOUBLE_EQUAL : TokenType.SINGLE_EQUAL);
            case '>' -> addType.accept(matchNext.apply('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
            case '<' -> addType.accept(matchNext.apply('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
            case ' ', '\r', '\t' -> {
            }
            case '"' -> matchString();
            case '\n' -> aLine++;
            default -> {
                if (isDigit.test(nextChar)) {
                    matchNumber();
                } else if (isAlpha.test(nextChar)) {
                    matchIdentifierAndReserved();
                } else {
                    JSurpher.error(aCurrent, "invalid character");
                }

            }
        }
    }

    /**
     * Appends a lexical token with a literal to the token list
     *
     * @param pType    the type of the lexical token
     * @param pLiteral the literal of the lexical token
     */
    private void addToken(TokenType pType, Object pLiteral) {
        aTokenList.add(new Token(pType, aSource.substring(aStart, aCurrent), pLiteral, aLine));
    }

    /**
     * Reads throught the source code and generates a list of lexical tokens
     *
     * @return a list of lexical tokens of Surpher
     */
    List<Token> scanTokens() {
        while (!isAtEnd.get()) {
            aStart = aCurrent;
            scanToken();
        }

        aTokenList.add(new Token(TokenType.EOF, "", null, aLine));
        return aTokenList;
    }
}

File: src/Lexer.cpp
#include <string>
#include <sstream>
#include <unordered_map>
#include <utility>

#include "Lexer.hpp"
#include "Token.hpp"
#include "Error.hpp"

static std::unordered_map<std::string, TokenType> keyWords = {
    {"class", CLASS},
    {"else", ELSE},
    {"false", FALSE},
    {"fun", FUN},
    {"for", FOR},
    {"if", IF},
    {"nil", NIL},
    {"return", RETURN},
    {"super", SUPER},
    {"this", THIS},
    {"true", TRUE},
    {"var", VAR},
    {"while", WHILE},
    {"do", DO},
    {"break", BREAK},
    {"continue", CONTINUE},
    {"print", PRINT},
    {"and", AND},
    {"or", OR},
    {"sig", SIG},
    {"import", IMPORT},
    {"namespace", NAMESPACE},
    {"fixed", FIXED},
    {"halt", HALT},
    {"alloc", ALLOC}};

Lexer::Lexer(std::string source_code) : source_code(std::move(source_code))
{
}

inline char Lexer::anyChar()
{
    return source_code[current++];
}

bool Lexer::matchNextChar(char expected)
{
    if (isAtEnd(0) || source_code[current] != expected)
        return false;
    current++;
    return true;
}

inline void Lexer::addToken(TokenType type, const std::any &literal)
{
    token_list.emplace_back(Token(source_code.substr(start, current - start), literal, type, line));
}

inline void Lexer::addToken(TokenType type)
{
    addToken(type, nullptr);
}

inline char Lexer::lookAHead(uint32_t offset)
{
    return (isAtEnd(0) || current + offset >= source_code.size()) ? '\0' : source_code[current + offset];
}

void Lexer::matchString()
{
    std::ostringstream str_builder;
    while (lookAHead(0) != '"' && !isAtEnd(0))
    {
        if (lookAHead(0) == '\n')
            line++;
        if (lookAHead(0) == '\\' && !isAtEnd(1))
        {
            switch (lookAHead(1))
            {
            case 't':
                str_builder << '\t';
                break;
            case 'n':
                str_builder << '\n';
                break;
            case 'b':
                str_builder << '\b';
                break;
            case 'f':
                str_builder << 'f';
                break;
            case '"':
                str_builder << '"';
                break;
            case '\\':
                str_builder << '\\';
                break;
            default:
                break;
            }
            anyChar();
            anyChar();
            continue;
        }
        str_builder << anyChar();
    }

    if (isAtEnd(0))
    {
        ::error(line, "Unterminated string.");
        return;
    }

    anyChar();

    TokenType type = STRING;
    std::any str_literal = str_builder.str();
    addToken(type, str_literal);
}

void Lexer::matchNumber()
{
    while (isDigit(lookAHead(0)) || lookAHead(0) == '.' || lookAHead(0) == 'e')
        anyChar();

    TokenType type = NUMBER;
    std::any num_literal;
    try
    {
        num_literal = std::stold(source_code.substr(start, current - start));
    }
    catch (const std::exception &e)
    {
        ::error(line, "Number format incorrect.");
    }

    addToken(type, num_literal);
}

void Lexer::matchIdentifierOrReserved()
{
    while (isAlphaNumeric(lookAHead(0)))
        anyChar();

    std::string word = source_code.substr(start, current - start);
    TokenType type;
    if (keyWords.find(word) != keyWords.end())
    {
        type = keyWords[word];
        addToken(type);
        return;
    }
    type = IDENTIFIER;
    addToken(type);
}

void Lexer::skipComment()
{
    uint32_t flag = 1;

    while (flag != 0 && !isAtEnd(0))
    {
        if (matchNextChar('/') && matchNextChar('*'))
        {
            flag++;
        }
        else if (matchNextChar('*') && matchNextChar('/'))
        {
            flag--;
            if (isAtEnd(0))
                return;
        }
        if (isAtEnd(0))
        {
            ::error(line, "Unterminated comment.");
            return;
        }
        else if (matchNextChar('\n'))
        {
            line++;
        }
        else
        {
            anyChar();
        }
    }
}

void Lexer::scanToken()
{
    char next_char = anyChar();
    switch (next_char)
    {
    case '(':
        addToken(LEFT_PAREN);
        break;
    case ')':
        addToken(RIGHT_PAREN);
        break;
    case '{':
        addToken(LEFT_BRACE);
        break;
    case '}':
        addToken(RIGHT_BRACE);
        break;
    case '[':
        addToken(LEFT_BRACKET);
        break;
    case '@':
        addToken(AT);
        break;
    case ']':
        addToken(RIGHT_BRACKET);
        break;
    case ',':
        addToken(COMMA);
        break;
    case '.':
        if (isDigit(lookAHead(0)))
        {
            matchNumber();
        }
        else
        {
            addToken(DOT);
        }
        break;
    case ';':
        addToken(SINGLE_SEMICOLON);
        break;
    case '+':
        addToken(PLUS);
        break;
    case '-':
        if (matchNextChar('>'))
        {
            addToken(RIGHT_ARROW);
        }
        else
        {
            addToken(MINUS);
        }
        break;
    case '?':
        addToken(QUESTION);
        break;
    case ':':
        addToken(SINGLE_COLON);
        break;
    case '%':
        addToken(PERCENT);
        break;
    case '*':
        addToken(STAR);
        break;
    case '\\':
        addToken(LAMBDA);
        break;
    case '/':
        if (matchNextChar('/'))
        {
            while (!isAtEnd(0) && !matchNextChar('\n'))
            {
                anyChar();
            }
            line++;
        }
        else if (matchNextChar('*'))
        {
            skipComment();
        }
        else
        {
            addToken(SLASH);
        }
        break;
    case '|':
        if (matchNextChar('>'))
        {
            addToken(PIPE);
        }
        else
        {
            addToken(SINGLE_BAR);
        }
        break;
    case '^':
        addToken(CARET);
        break;
    case '&':
        addToken(SINGLE_AMPERSAND);
        break;
    case '!':
        if (matchNextChar('='))
        {
            addToken(BANG_EQUAL);
        }
        else
        {
            addToken(BANG);
        }
        break;
    case '=':
        if (matchNextChar('='))
        {
            addToken(DOUBLE_EQUAL);
        }
        else
        {
            addToken(SINGLE_EQUAL);
        }
        break;
    case '>':
        if (matchNextChar('='))
        {
            addToken(GREATER_EQUAL);
        }
        else if (matchNextChar('>'))
        {
            addToken(RIGHT_SHIFT);
        }
        else
        {
            addToken(GREATER);
        }
        break;
    case '<':
        if (matchNextChar('='))
        {
            addToken(LESS_EQUAL);
        }
        else if (matchNextChar('<'))
        {
            addToken(LEFT_SHIFT);
        }
        else
        {
            addToken(LESS);
        }
        break;
    case ' ':
    case '\r':
    case '\t':
        break;
    case '"':
        matchString();
        break;
    case '\n':
        line++;
        break;
    default:
        if (isDigit(next_char))
        {
            matchNumber();
        }
        else if (isAlpha(next_char))
        {
            matchIdentifierOrReserved();
        }
        else
        {
            ::error(current, "invalid character");
        }
    }
}

std::vector<Token> Lexer::scanTokens()
{
    while (!isAtEnd(0))
    {
        start = current;
        scanToken();
    }

    TokenType type = EOF_TOKEN;
    token_list.emplace_back(Token("", nullptr, type, line));
    return token_list;
}

inline bool Lexer::isAtEnd(uint32_t offset)
{
    return current + offset >= source_code.size();
}

inline bool Lexer::isDigit(char c)
{
    return c >= '0' && c <= '9';
}

inline bool Lexer::isAlpha(char c)
{
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c == '_');
}

inline bool Lexer::isAlphaNumeric(char c)
{
    return isDigit(c) || isAlpha(c);
}
